### DISCUSSION TITLE
[RFC] 052 - Cloudflare Workers AI Support

### DISCUSSION THREAD

**Discussion by BrandonStudio** [lobehub/lobehub#3186] [RFCs]
# Add [Cloudflare Workers AI](https://developers.cloudflare.com/workers-ai) provider.

[Cloudflare Workers AI](https://developers.cloudflare.com/workers-ai/models) now supports following types of models:
- [**Text Generation**](https://developers.cloudflare.com/workers-ai/models#text-generation)
- [**Image-to-Text**](https://developers.cloudflare.com/workers-ai/models#image-to-text)
- [**Automatic Speech Recognition**](https://developers.cloudflare.com/workers-ai/models#automatic-speech-recognition)
- [**Translation**](https://developers.cloudflare.com/workers-ai/models#translation)
- [**Text-to-Image**](https://developers.cloudflare.com/workers-ai/models#text-to-image)
- [**Text Embeddings**](https://developers.cloudflare.com/workers-ai/models#text-embeddings)
- [Image Classification](https://developers.cloudflare.com/workers-ai/models#image-classification)
- [Object Detection](https://developers.cloudflare.com/workers-ai/models#object-detection)
- [Summarization](https://developers.cloudflare.com/workers-ai/models#summarization)
- [Text Classification](https://developers.cloudflare.com/workers-ai/models#text-classification)

Where the first 5 types of models can be used in LobeChat.

- [ ] Use text generation models as chat models.
       Done in https://github.com/lobehub/lobe-chat/pull/3402, awaiting https://github.com/lobehub/lobe-chat/pull/4671.
- [ ] Use Image-to-Text models as vision chat models.
- [ ] Use automatic speech recognition models as speech recognizer.
- [ ] Use translation models as translator (system assistant).
- [ ] Similar to DALL-E, add built-in plugin for text-to-image models.
- [ ] Use embedding models to help build knowledge base.

## Required configurations
- Account ID
  This can be replaced with endpoint url.
- Authorization
  This can be either Bearer token (`Authorization: Bearer {token}`) or API key (`X-Auth-Key: {key}`).
- Endpoint URL
  This allows user to use AI Gateway or their own workers, which may be better than account ID.

## Chat model capability
- ✔ All models support stream.
- ✖ Few models (one for now) support function calling.

## Compatibility
- Although Cloudflare Workers AI is compatible with OpenAI API endpoints, it does not support function calling there.
- Users may have created a worker to reversely proxy AI requests. If this use case is permitted, authorization config could be optional.

**Comment by arvinxx** [comment]
there is a PR for it : #2966

**Reply by BrandonStudio** [reply]
I see it now. It would be great, but
1. I think that PR uses OpenAI style API, which does not support function calling.
2. I think that PR does not support both bearer auth and api token.
3. It does not supports a custom endpoint.

**Comment by sxjeru** [comment]
Thank you for your proposal. But sorry, due to my technical limitations, I only consider implementing basic text generation. 
I hope other contributors can complete and improve other functions.

**Reply by BrandonStudio** [reply]
Do you think we should support custom endpoint? What do you think of the difference between using account ID and custom endpoint?

**Reply by sxjeru** [reply]
Custom endpoints can be considered. But I don't think there will be many people using it.

**Comment by BrandonStudio** [comment]
@arvinxx How to handle an event-stream response?
Streams from Cloudflare Workers AI is like:
```
data: {"response":"你好","p":"abcdefghijklmnop"}
data: {"response":"！","p":"abcdefghijklmnopqrstuvwxyz012345"}
data: {"response":"很高兴","p":"abcdefghijklmnopqrstuv"}
data: {"response":"为你","p":"abcdefghijklmnopqrstuvwxyz0123456789abcdefghij"}
data: {"response":"提供","p":"abcdefghijklmnopqrstuvwxyz0123456789abcdefghijk"}
data: {"response":"帮助","p":"abc"}
data: {"response":"。","p":"a"}
data: {"response":"","p":"abcdefghij"}
data: [DONE]
```
Parsed:
```
你好！很高兴为你提供帮助。
```
And tool call response is like
```
data: {"response":"<","p":"abcdefghijk"}
data: {"response":"tool","p":"ab"}
data: {"response":"_","p":"abcdefghijklmnopqrstuvwxyz0123456789abc"}
data: {"response":"call","p":"abcdefghijklmn"}
data: {"response":">","p":"abcdefghijklmnopqrstuvwxyz0123456789ab"}
data: {"response":"\n","p":"abcdefghijklmnopqrstuvwxyz0123456"}
data: {"response":"{'","p":"abcdefghijklmnopqrstuvwxyz0123456789ab"}
data: {"response":"arguments","p":"abcdefghijklmnopqrstuvwxyz0123456789abcdefghijklm"}
data: {"response":"':","p":"abcdefghijklmnop"}
data: {"response":" {'","p":"abcdefghijklmnopqrstuvwxyz0123456789abcdefghijklm"}
data: {"response":"username","p":"abcdefghijklmnop"}
data: {"response":"':","p":"abcdefghijklmnopqrstuv"}
data: {"response":" '","p":"abcdefghijklmnopqrstuvwxy"}
data: {"response":"cloud","p":"abcdefghijklmnopqrstuvwxyz0123456"}
data: {"response":"fl","p":"abcdefghijklmnopqrstuvwxyz012345"}
data: {"response":"are","p":"abcdefghijklmnopqrstuvwxyz012345678"}
data: {"response":"'},","p":"abcdefghijklmnopqrstuvwxyz0123"}
data: {"response":" '","p":"abcdefghijklmnopqrstuvwxyz0123456789abcd"}
data: {"response":"name","p":"abcdefghijkl"}
data: {"response":"':","p":"abcdefghijklmnopqrstuvwxy"}
data: {"response":" '","p":"abcdefghijklmno"}
data: {"response":"get","p":"abcdefghijklmnopqrstuvwxyz0"}
data: {"response":"G","p":"abcdefghijklmnopqrstuvwxyz01234"}
data: {"response":"ithub","p":"ab"}
data: {"response":"User","p":"abcdefghijklmnopqrs"}
data: {"response":"'}","p":"abcdefghijklmnopqrstuvwxyz012"}
data: {"response":"\n"}
data: {"response":"</","p":"abcdefghijk"}
data: {"response":"tool","p":"abcdefghijklmnopqrstuvwxyz0123456789abcdefg"}
data: {"response":"_","p":"abcdefghijklmnopq"}
data: {"response":"call","p":"abcdefghijklmnopqrstuvwxyz0123456789abcdef"}
data: {"response":">","p":"abcdefghijklmnopqrstuvwxyz0123456789a"}
data: {"response":"<|im_end|>","p":"abcdefghijklmnopqrstuvwxyz012345678"}
data: [DONE]
```
Parsed:
```
<tool_call>\n{'arguments': {'username': 'cloudflare'}, 'name': 'getGithubUser'}\n</tool_call><|im_end|>
```
The stream response of a tool call is not a JSON.
However, non-stream response is more readable:
```json
{"result":{"response":null,"tool_calls":[{"arguments":{"username":"cloudflare"},"name":"getGithubUser"}]},"success":true,"errors":[],"messages":[]}
```

**Reply by arvinxx** [reply]
here is an example of handle stream response: https://github.com/lobehub/lobe-chat/discussions/2794

**Comment by arvinxx** [comment]
@BrandonStudio I will write a post to talk about streaming tools calling these day. Maybe it will help.

**Reply by BrandonStudio** [reply]
By the way, how to integrate dynamic model list? Do we have a refresh delegate by which user will get up-to-date model list at front end? If we do, where should I set the handler?
As at [BrandonStudio#1/index.ts#L163](https://github.com/BrandonStudio/lobe-chat/blob/ebab075af75dde255a73a5961bf225cb5d081e24/src/libs/agent-runtime/cloudflare/index.ts#L163), I have a `getModels` method. How could I integrate it to the project?

**Reply by arvinxx** [reply]
There should be a param to auto fetch models when app init. But currently we don't implement it. If implemented, it should be as easy as to add a simple field like `autoFetchModelList: true` to `ModelCard`.

**Reply by BrandonStudio** [reply]
I saw `models` method defination in `LobeRuntimeAI`, and `OpenAICompatibleFactory` has already implemented it.
But when I rename `getModels` to `models`, although the response is in correct format, and `data` in [`onSuccess`](https://github.com/lobehub/lobe-chat/blob/ea798e57d4ad5be40953d8586e2daabe1d1a48be/src/store/user/slices/modelList/action.ts#L198) **does** appear to be an array, the pull-down list **keeps** unchanged.

![image](https://github.com/user-attachments/assets/55cb2781-40e1-40ff-9831-e87e3524e16b)

![image](https://github.com/user-attachments/assets/ba9061a3-51ab-457e-896f-fd88371b9fbf)

I'm confused now. Do you mean that this delegate is not yet used to update the list?

**Reply by BrandonStudio** [reply]
I know now. I should add configuration in [`defaultModelProviderList`](https://github.com/lobehub/lobe-chat/blob/ea798e57d4ad5be40953d8586e2daabe1d1a48be/src/store/user/slices/modelList/action.ts#L91-L109) closure.
